ssl/ssl_ciph.c

```diff
@@ -1749,6 +1749,13 @@ uint32_t SSL_CIPHER_get_id(const SSL_CIPHER *c)
     return c->id;
 }
 
+uint32_t SSL_CIPHER_get_standard_id(const SSL_CIPHER *c)
+{
+    if (c->id & SSL3_CK_CIPHERSUITE_FLAG)
+        return c->id & ~SSL3_CK_CIPHERSUITE_FLAG;
+    return c->id;
```

Comment:
This function body leaves me a little puzzled.  It seems that if SSL3_CK_CIPHERSUITE_FLAG is not set, then the SSL_CIPHER object is unlikely to correspond to an IANA-allocated cipher codepoint, and we might be better off returning a specific error sentinel (0 is an allocated value for NULL_NULL_WITH_NULL, unfortunately, so it may not be the best candidate).
In any case, it seems quite difficult to obtain an SSL_CIPHER object with an id that does not have the SSL3_CK_CIPHERSUITE_FLAG set, as even the elements of ssl3_scsvs contain it.  So perhaps the question is moot, but I would like to hear a bit more from Paul about why this choice was made.

Reply:
Ahhh, the reason was that I was not very sure about if all cipher IDs get masked by SSL3_CK_CIPHERSUITE_FLAG or not (now and in future). If some ciphers introduced in the future (say TLS v1.4?) don't have any masks/prefixes, this function could also work fine (Just as if SSLv2's ciphers, though they are gone now). I admire this is a 'lazy' thought: to just return the original ID that we haven't covered in the code yet.

Returning an error value, I am also not sure since we need extra effort to define such a value. So how about define this `SSL_CIPHER_get_standard_id` function as: if the cipher has an IANA-specific ID, then returns it; otherwise returns its OpenSSL-specific IDs.

Reply:
All ciphers in the tree currently have the 0x3000 prefix, including the GOST ciphers (that are not listed in the IANA registry!).  I am not worried about any hypothetical future changes breaking this invariant out from under this code -- such changes would be quite sweeping and would need to look all over the tree, and should be expected to also find this code.

The GOST examples make your proposed documentation wording/definition troublesome; maybe we should refer to the wire protocol (treated as a 16-bit unsigned integer) instead, in which case it might be simpler to just return ```c->id & 0xffff```.

Reply:
Referring to wire protocol is a good idea. I will apply this.